File: src/tflite_detector.py
from pathlib import Path

import cv2
import numpy as np

# Import TensorFlow Lite lazily so that code importing this
# module still works where TFLite is not installed.

try:
    from tflite_runtime.interpreter import Interpreter
except ImportError:
    try:
        from tensorflow.lite.python.interpreter import Interpreter
    except ImportError:
        Interpreter = None


def letterbox(image, size, fill_value=114):
    """
    Resize an image to a square of `size` keeping the aspect ratio.

    Returns:
        (canvas, scale, left, top)
        scale: applied resize factor
        left, top: pad offsets in the letterboxed image
    """

    height, width = image.shape[:2]

    scale = min(
        size / width,
        size / height
    )

    new_width = max(
        int(round(width * scale)),
        1
    )

    new_height = max(
        int(round(height * scale)),
        1
    )

    resized = cv2.resize(
        image,
        (new_width, new_height),
        interpolation=cv2.INTER_LINEAR
    )

    left = (size - new_width) // 2
    top = (size - new_height) // 2

    canvas = np.full(
        (size, size, 3),
        fill_value=fill_value,
        dtype=image.dtype
    )

    canvas[
        top:top + new_height,
        left:left + new_width
    ] = resized

    return canvas, scale, left, top


class TFLiteDetector:
    """
    Person detection using a TensorFlow Lite model.

    Supports YOLO-style heads (single raw output tensor)
    and post-processed heads with built-in NMS
    (boxes, classes, scores, num_detections).

    Provides the same detect_person() interface as YOLODetector.
    """

    PERSON_CLASS_ID = 0

    def __init__(self, model_path, threads=4):
        if Interpreter is None:
            raise ImportError(
                "TensorFlow Lite is not installed. "
                "Install 'tflite-runtime' on the Raspberry Pi."
            )

        self.model_path = Path(model_path)

        if not self.model_path.exists():

            raise FileNotFoundError(
                f"Model file not found: {self.model_path}"
            )

        self._interpreter = Interpreter(
            model_path=str(self.model_path),
            num_threads=int(threads)
        )

        self._interpreter.allocate_tensors()

        self._input_details = (
            self._interpreter.get_input_details()
        )

        self._output_details = (
            self._interpreter.get_output_details()
        )

        input_detail = self._input_details[0]

        self._input_index = input_detail["index"]

        self._input_shape = (
            input_detail["shape"]
        )

        # Fixed square inference size
        # [1, size, size, 3] or [1, 3, size, size]

        self._inference_size = int(self._input_shape[1])

        self._input_dtype = np.dtype(
            input_detail["dtype"]
        )

        if self._input_dtype == np.uint8:

            # Quantized input: feed pixels 0..255 as-is.
            self._scale_bytes = True

        else:

            # Float input: divide by 255.
            self._scale_bytes = False

        self._prepare_output_details()

        print(
            f"TFLite model loaded: {self.model_path.name}"
        )

        print(
            f"Input shape: {self._input_shape}"
        )


    def _prepare_output_details(self):
        """
        Identify the detection output layout.

        Two layouts are supported:

        1. YOLO-style raw head (single tensor [1, 84, N]).
        2. Post-processed TFLite detection head (built-in NMS).
           The standard published order is:

               output 0: detection_boxes    (1, N, 4)
               output 1: detection_classes  (1, N)
               output 2: detection_scores   (1, N)
               output 3: num_detections     (1)
        """

        self._yolo_head = False
        self._yolo_index = None

        self._boxes_index = 0
        self._classes_index = 1
        self._scores_index = 2
        self._num_index = None

        num_outputs = len(self._output_details)

        if num_outputs > 3:
            self._num_index = 3

        for index, detail in enumerate(
            self._output_details
        ):

            shape = tuple(detail["shape"])

            # YOLO head: [1, 84, N].
            if (
                len(shape) == 3
                and shape[1] == 84
            ):
                self._yolo_head = True
                self._yolo_index = index
                return

            # Post-processed head: boxes end with a 4.
            if (
                len(shape) == 3
                and shape[2] == 4
            ):
                self._boxes_index = index


    def detect_person(
        self,
        frame,
        confidence_threshold=0.5
    ):
        """
        Detect the most confident person.

        Returns:
            Same dictionary format as YOLODetector:
            class_name, confidence, x1, y1, x2, y2

            or None
        """

        blob, scale, left, top = (
            self._preprocess(frame)
        )

        outputs = self._run(blob)

        if self._yolo_head:

            return self._parse_yolo(
                outputs,
                scale, left, top,
                confidence_threshold
            )

        return self._parse_postprocessed(
            outputs,
            scale, left, top,
            confidence_threshold
        )


    def _preprocess(self, frame):
        """Convert frame to the model input tensor."""

        # Models are trained on RGB images.

        rgb = cv2.cvtColor(
            frame, cv2.COLOR_BGR2RGB
        )

        blob, scale, left, top = letterbox(
            rgb,
            self._inference_size
        )

        # Add batch dimension -> (1, size, size, 3)

        blob = np.expand_dims(
            blob,
            axis=0
        )

        if not self._scale_bytes:

            blob = blob.astype(np.float32) / 255.0

        return blob, scale, left, top


    def _run(self, blob):
        """Run inference and return all outputs."""

        self._interpreter.set_tensor(
            self._input_index,
            blob
        )

        self._interpreter.invoke()

        return [
            self._interpreter.get_tensor(
                detail["index"]
            )
            for detail in self._output_details
        ]


    def _map_box(self, x1, y1, x2, y2, scale, left, top):
        """
        Map box coordinates from model input to original frame.
        """

        x1 = (x1 - left) / scale
        y1 = (y1 - top) / scale
        x2 = (x2 - left) / scale
        y2 = (y2 - top) / scale

        return int(x1), int(y1), int(x2), int(y2)


    def _parse_yolo(
        self,
        outputs,
        scale, left, top,
        confidence_threshold
    ):
        """
        Parse a YOLO-style head (single tensor [1, 84, N]).
        """

        predictions = outputs[self._yolo_index][0]

        predictions = predictions.T

        best_detection = None

        best_confidence = confidence_threshold


        for prediction in predictions:

            x_center = prediction[0]
            y_center = prediction[1]

            width = prediction[2]
            height = prediction[3]

            confidence = prediction[
                4 + self.PERSON_CLASS_ID
            ]

            if confidence < best_confidence:
                continue

            best_confidence = float(confidence)

            x1 = x_center - width / 2
            y1 = y_center - height / 2

            x2 = x_center + width / 2
            y2 = y_center + height / 2

            x1, y1, x2, y2 = self._map_box(
                x1, y1, x2, y2,
                scale, left, top
            )

            best_detection = {
                "class_name": "person",
                "confidence": best_confidence,
                "x1": x1,
                "y1": y1,
                "x2": x2,
                "y2": y2
            }

        return best_detection


    def _parse_postprocessed(
        self,
        outputs,
        scale, left, top,
        confidence_threshold
    ):
        """
        Parse a post-processed head with built-in NMS
        (SSD / EfficientDet style).
        """

        boxes = outputs[self._boxes_index][0]

        scores = outputs[self._scores_index][0]

        classes = outputs[self._classes_index][0]

        frame_size = self._inference_size

        num_detections = len(scores)

        if self._num_index is not None:

            detected = int(outputs[self._num_index][0])

            if detected > 0:
                num_detections = min(detected, num_detections)

        # The built-in NMS returns the strongest detections first,
        # so class 0 (person) within the top detections wins.
        # Scores below the threshold are ignored.

        for i in range(num_detections):

            class_id = int(classes[i])

            confidence = float(scores[i])

            if class_id != self.PERSON_CLASS_ID:
                continue

            if confidence < confidence_threshold:
                continue

            ymin, xmin, ymax, xmax = boxes[i]

            x1, y1, x2, y2 = self._map_box(
                xmin * frame_size,
                ymin * frame_size,
                xmax * frame_size,
                ymax * frame_size,
                scale, left, top
            )

            return {
                "class_name": "person",
                "confidence": confidence,
                "x1": x1,
                "y1": y1,
                "x2": x2,
                "y2": y2
            }

        return None

File: cpp/src/tflite_detector.cpp
#include "tflite_detector.hpp"

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <filesystem>
#include <iostream>
#include <stdexcept>

#include <opencv2/imgproc.hpp>

namespace tp {

namespace {

/**
 * Resize an image to a square of `size` keeping the aspect ratio.
 * Mirrors letterbox() in src/tflite_detector.py (fill value 114).
 */
cv::Mat letterbox(const cv::Mat& image, int size, double& scale,
                  int& left, int& top) {
    const int width = image.cols;
    const int height = image.rows;

    scale = std::min(static_cast<double>(size) / width,
                     static_cast<double>(size) / height);

    const int new_width = std::max(static_cast<int>(std::round(width * scale)), 1);
    const int new_height = std::max(static_cast<int>(std::round(height * scale)), 1);

    cv::Mat resized;
    cv::resize(image, resized, cv::Size(new_width, new_height), 0, 0,
               cv::INTER_LINEAR);

    left = (size - new_width) / 2;
    top = (size - new_height) / 2;

    cv::Mat canvas(size, size, image.type(), cv::Scalar(114, 114, 114));
    resized.copyTo(canvas(cv::Rect(left, top, new_width, new_height)));
    return canvas;
}

#ifdef TP_HAVE_TFLITE
std::vector<float> tensorToFloat(const TfLiteTensor* tensor) {
    int count = 1;
    for (int i = 0; i < tensor->dims->size; ++i) {
        count *= tensor->dims->data[i];
    }

    std::vector<float> out(static_cast<std::size_t>(count));

    if (tensor->type == kTfLiteFloat32) {
        const float* data = tensor->data.f;
        out.assign(data, data + count);
    } else if (tensor->type == kTfLiteUInt8) {
        const uint8_t* data = tensor->data.uint8;
        const float scale = tensor->params.scale;
        const float zero_point = static_cast<float>(tensor->params.zero_point);
        for (int i = 0; i < count; ++i) {
            out[static_cast<std::size_t>(i)] = (static_cast<float>(data[i]) - zero_point) * scale;
        }
    } else if (tensor->type == kTfLiteInt8) {
        const int8_t* data = tensor->data.int8;
        const float scale = tensor->params.scale;
        const float zero_point = static_cast<float>(tensor->params.zero_point);
        for (int i = 0; i < count; ++i) {
            out[static_cast<std::size_t>(i)] = (static_cast<float>(data[i]) - zero_point) * scale;
        }
    } else if (tensor->type == kTfLiteInt32) {
        const int32_t* data = tensor->data.i32;
        for (int i = 0; i < count; ++i) {
            out[static_cast<std::size_t>(i)] = static_cast<float>(data[i]);
        }
    }

    return out;
}
#endif  // TP_HAVE_TFLITE

}  // namespace

TFLiteDetector::TFLiteDetector(const std::string& model_path, int threads)
    : model_path_(model_path) {
#ifdef TP_HAVE_TFLITE
    if (!std::filesystem::exists(model_path_)) {
        throw std::runtime_error("Model file not found: " + model_path_);
    }

    model_ = tflite::FlatBufferModel::BuildFromFile(model_path_.c_str());
    if (!model_) {
        throw std::runtime_error("Cannot build TFLite model from file: " + model_path_);
    }

    tflite::ops::builtin::BuiltinOpResolver resolver;
    tflite::InterpreterBuilder builder(*model_, resolver);
    if (builder(&interpreter_) != kTfLiteOk || !interpreter_) {
        throw std::runtime_error("Cannot build TFLite interpreter for: " + model_path_);
    }

    if (threads > 0) {
        interpreter_->SetNumThreads(threads);
    }
    if (interpreter_->AllocateTensors() != kTfLiteOk) {
        throw std::runtime_error("Cannot allocate TFLite tensors for: " + model_path_);
    }

    const int input_index = interpreter_->inputs()[0];
    const TfLiteTensor* input = interpreter_->tensor(input_index);
    const int* dims = input->dims->data;

    // Fixed square inference size ([1, size, size, 3]).
    inference_size_ = dims[1];
    scale_bytes_ = (input->type == kTfLiteUInt8) || (input->type == kTfLiteInt8);

    prepareOutputDetails();

    std::cout << "TFLite model loaded: "
              << std::filesystem::path(model_path_).filename().string()
              << std::endl;
    std::cout << "Input shape: [";
    for (int i = 0; i < input->dims->size; ++i) {
        std::cout << (i == 0 ? "" : ", ") << dims[i];
    }
    std::cout << "]" << std::endl;
    std::cout << "Input type: "
              << (scale_bytes_ ? "int8 (quantized)" : "float32")
              << std::endl;
#else
    (void)threads;
    throw std::runtime_error(
        "TensorFlow Lite support not compiled in (TP_HAVE_TFLITE). Rebuild "
        "with TP_ENABLE_TFLITE=ON and a working TensorFlow Lite install.");
#endif
}

void TFLiteDetector::prepareOutputDetails() {
#ifdef TP_HAVE_TFLITE
    yolo_head_ = false;
    yolo_index_ = 0;
    boxes_index_ = 0;
    classes_index_ = 1;
    scores_index_ = 2;
    num_index_ = -1;

    const int num_outputs = static_cast<int>(interpreter_->outputs().size());
    if (num_outputs > 3) {
        num_index_ = 3;
    }

    for (int index = 0; index < num_outputs; ++index) {
        const TfLiteTensor* output = interpreter_->tensor(interpreter_->outputs()[index]);
        const int rank = output->dims->size;

        // YOLO head: [1, 84, N].
        if (rank == 3 && output->dims->data[1] == 84) {
            yolo_head_ = true;
            yolo_index_ = index;
            return;
        }

        // Post-processed head: boxes end with a 4.
        if (rank == 3 && output->dims->data[2] == 4) {
            boxes_index_ = index;
        }
    }
#else
    yolo_head_ = false;
#endif
}

void TFLiteDetector::preprocess(const cv::Mat& frame, double& scale,
                                int& left, int& top) {
#ifdef TP_HAVE_TFLITE
    // Models are trained on RGB images.
    cv::Mat rgb;
    cv::cvtColor(frame, rgb, cv::COLOR_BGR2RGB);

    cv::Mat canvas = letterbox(rgb, inference_size_, scale, left, top);

    const TfLiteTensor* input = interpreter_->tensor(interpreter_->inputs()[0]);
    const std::size_t count = static_cast<std::size_t>(inference_size_) *
                              inference_size_ * 3;
    const uint8_t* src = canvas.data;

    if (scale_bytes_) {
        if (input->type == kTfLiteInt8) {
            int8_t* data = input->data.int8;
            for (std::size_t i = 0; i < count; ++i) {
                data[i] = static_cast<int8_t>(static_cast<int>(src[i]) - 128);
            }
        } else {
            uint8_t* data = input->data.uint8;
            std::memcpy(data, src, count);
        }
    } else {
        float* data = input->data.f;
        const float inv = 1.0f / 255.0f;
        for (std::size_t i = 0; i < count; ++i) {
            data[i] = src[i] * inv;
        }
    }
#endif
}

bool TFLiteDetector::invoke() {
#ifdef TP_HAVE_TFLITE
    return interpreter_->Invoke() == kTfLiteOk;
#else
    return false;
#endif
}

Detection TFLiteDetector::detectPerson(const cv::Mat& frame,
                                       float confidence_threshold) {
#ifdef TP_HAVE_TFLITE
    if (!frame.data) {
        return Detection{};
    }

    double scale;
    int left = 0, top = 0;
    preprocess(frame, scale, left, top);

    if (!invoke()) {
        return Detection{};
    }

    if (yolo_head_) {
        return parseYolo(scale, left, top, confidence_threshold);
    }
    return parsePostprocessed(scale, left, top, confidence_threshold);
#else
    (void)frame;
    (void)confidence_threshold;
    return Detection{};
#endif
}

Detection TFLiteDetector::mapBox(double x1, double y1, double x2, double y2,
                                 double scale, int left, int top) {
    Detection detection;
    detection.x1 = static_cast<int>((x1 - left) / scale);
    detection.y1 = static_cast<int>((y1 - top) / scale);
    detection.x2 = static_cast<int>((x2 - left) / scale);
    detection.y2 = static_cast<int>((y2 - top) / scale);
    return detection;
}

Detection TFLiteDetector::parseYolo(double scale, int left, int top,
                                    float confidence_threshold) {
#ifdef TP_HAVE_TFLITE
    const TfLiteTensor* tensor = interpreter_->tensor(interpreter_->outputs()[yolo_index_]);
    const int num_anchors = tensor->dims->data[2];
    const std::vector<float> predictions = tensorToFloat(tensor);

    Detection best;
    float best_confidence = confidence_threshold;

    for (int n = 0; n < num_anchors; ++n) {
        auto at = [&](int k) -> float {
            return predictions[static_cast<std::size_t>(k) * num_anchors + n];
        };

        const float x_center = at(0);
        const float y_center = at(1);
        const float width = at(2);
        const float height = at(3);
        const float confidence = at(4 + PERSON_CLASS_ID);

        if (confidence < best_confidence) {
            continue;
        }
        best_confidence = confidence;

        best = mapBox(x_center - width / 2, y_center - height / 2,
                      x_center + width / 2, y_center + height / 2,
                      scale, left, top);
        best.class_name = "person";
        best.confidence = confidence;
    }

    return best;
#else
    (void)scale;
    (void)left;
    (void)top;
    (void)confidence_threshold;
    return Detection{};
#endif
}

Detection TFLiteDetector::parsePostprocessed(double scale, int left, int top,
                                             float confidence_threshold) {
#ifdef TP_HAVE_TFLITE
    const std::vector<float> boxes =
        tensorToFloat(interpreter_->tensor(interpreter_->outputs()[boxes_index_]));
    const std::vector<float> scores =
        tensorToFloat(interpreter_->tensor(interpreter_->outputs()[scores_index_]));
    const std::vector<float> classes =
        tensorToFloat(interpreter_->tensor(interpreter_->outputs()[classes_index_]));

    const int max_detections = static_cast<int>(scores.size());
    int num_detections = max_detections;
    if (num_index_ >= 0) {
        const std::vector<float> counts =
            tensorToFloat(interpreter_->tensor(interpreter_->outputs()[num_index_]));
        const int detected = counts.empty() ? 0 : static_cast<int>(counts[0]);
        if (detected > 0) {
            num_detections = std::min(detected, max_detections);
        }
    }

    const int frame_size = inference_size_;
    for (int i = 0; i < num_detections; ++i) {
        const int class_id = static_cast<int>(classes[i]);
        const float confidence = scores[i];
        if (class_id != PERSON_CLASS_ID) {
            continue;
        }
        if (confidence < confidence_threshold) {
            continue;
        }

        const double ymin = boxes[i * 4 + 0];
        const double xmin = boxes[i * 4 + 1];
        const double ymax = boxes[i * 4 + 2];
        const double xmax = boxes[i * 4 + 3];

        Detection detection = mapBox(xmin * frame_size, ymin * frame_size,
                                     xmax * frame_size, ymax * frame_size,
                                     scale, left, top);
        detection.class_name = "person";
        detection.confidence = confidence;
        return detection;
    }

    return Detection{};
#else
    (void)scale;
    (void)left;
    (void)top;
    (void)confidence_threshold;
    return Detection{};
#endif
}

}  // namespace tp